File: backend/src/models/subscriber.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Subscriber {
    pub id: Uuid,
    pub user_id: Uuid,
    pub email: String,
    pub name: Option<String>,
    pub status: SubscriberStatus,
    pub tags: Vec<String>,
    pub custom_fields: serde_json::Value,
    pub subscribed_at: DateTime<Utc>,
    pub unsubscribed_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, sqlx::Type, Clone, Copy)]
#[sqlx(type_name = "subscriber_status", rename_all = "lowercase")]
pub enum SubscriberStatus {
    Active,
    Unsubscribed,
    Bounced,
    Complained,
}

// 購読者作成リクエスト
#[derive(Debug, Serialize, Deserialize, Validate)]
pub struct CreateSubscriberRequest {
    #[validate(email(message = "有効なメールアドレスを入力してください"))]
    pub email: String,
    pub name: Option<String>,
    pub status: Option<SubscriberStatus>,
    pub tags: Option<Vec<String>>,
    pub custom_fields: Option<serde_json::Value>,
}

// 購読者更新リクエスト
#[derive(Debug, Serialize, Deserialize, Validate)]
pub struct UpdateSubscriberRequest {
    #[validate(email(message = "有効なメールアドレスを入力してください"))]
    pub email: Option<String>,
    pub name: Option<String>,
    pub status: Option<SubscriberStatus>,
    pub tags: Option<Vec<String>>,
    pub custom_fields: Option<serde_json::Value>,
}

// CSVインポートリクエスト
#[derive(Debug, Serialize, Deserialize)]
pub struct ImportSubscribersRequest {
    pub csv_content: String,
    pub has_header: Option<bool>,
    pub column_mapping: ColumnMapping,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ColumnMapping {
    pub email: usize,
    pub name: Option<usize>,
    pub tags: Option<Vec<usize>>,
    pub custom_fields: Option<Vec<CustomFieldMapping>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CustomFieldMapping {
    pub name: String,
    pub column: usize,
}

// 購読者一覧リクエストオプション
#[derive(Debug, Serialize, Deserialize, Default)]
pub struct ListSubscriberOptions {
    pub limit: Option<i64>,
    pub offset: Option<i64>,
    pub search: Option<String>,
    pub tag: Option<String>,
    pub status: Option<String>,
    pub sort_by: Option<String>,
    pub sort_order: Option<String>,
}

// 購読者一覧レスポンス
#[derive(Debug, Serialize, Deserialize)]
pub struct SubscriberListResponse {
    pub subscribers: Vec<Subscriber>,
    pub total: i64,
    pub available_tags: Vec<String>,
}

// インポート結果レスポンス
#[derive(Debug, Serialize, Deserialize)]
pub struct ImportSubscribersResponse {
    pub imported_count: u32,
    pub errors: Vec<String>,
}

File: frontend/src/lib/types/subscriber.ts
// import type { User } from "../stores/authStore";

/**
 * 購読者のステータス
 */
export enum SubscriberStatus {
  ACTIVE = "active",
  UNSUBSCRIBED = "unsubscribed",
  BOUNCED = "bounced",
  COMPLAINED = "complained",
}

/**
 * 購読者モデル
 */
export interface Subscriber {
  id: string;
  user_id: string;
  email: string;
  name?: string;
  status: SubscriberStatus;
  tags: string[];
  custom_fields: Record<string, unknown>;
  subscribed_at: string;
  unsubscribed_at?: string;
  created_at: string;
  updated_at: string;
}

/**
 * 購読者作成リクエスト
 */
export interface CreateSubscriberRequest {
  email: string;
  name?: string;
  tags?: string[];
  custom_fields?: Record<string, unknown>;
}

/**
 * 購読者更新リクエスト
 */
export interface UpdateSubscriberRequest {
  name?: string;
  status?: SubscriberStatus;
  tags?: string[];
  custom_fields?: Record<string, unknown>;
}

/**
 * 購読者CSV一括インポートリクエスト
 */
export interface ImportSubscribersRequest {
  file: unknown; // File type from DOM API
  tag?: string;
}

/**
 * 購読者一覧レスポンス
 */
export interface SubscriberListResponse {
  subscribers: Subscriber[];
  total: number;
  limit: number;
  offset: number;
  available_tags?: string[];
}

/**
 * 購読者インポートレスポンス
 */
export interface ImportSubscribersResponse {
  message: string;
  imported: number;
  failed: number;
  errors?: string[];
}
